ui/src/app/views/profile-page/component.tsx

```diff
@@ -0,0 +1,314 @@
+import {SpinnerOverlay} from 'app/components/spinners';
+import {profileApi} from 'app/services/swagger-fetch-clients';
+import colors from 'app/styles/colors';
+import {reactStyles, ReactWrapperBase, withUserProfile} from 'app/utils';
+import {Profile} from 'generated/fetch';
+
+
+const styles = reactStyles({
+  h1: {
+    color: colors.purple[0],
+    fontSize: 20,
+    fontWeight: 500,
+    lineHeight: '24px'
+  },
+  inputLabel: {
+    color: colors.purple[0],
+    fontSize: 14, lineHeight: '18px',
+    marginBottom: 6
+  },
+  inputStyle: {
+    width: 250,
+    marginRight: 20
+  },
+  longInputStyle: {
+    height: 175, width: 420,
+    resize: 'both'
+  },
+  box: {
+    backgroundColor: '#fff',
+    borderRadius: 8,
+    padding: 21
+  }
+});
+
+const labelSubstitutions = {
+  givenName: 'firstName',
+  familyName: 'lastName',
+  currentPosition: 'currentPosition',
+  organization: 'organization',
+  areaOfResearch: 'currentResearch'
+};
+const required = {presence: {allowEmpty: false}};
+const notTooLong = maxLength => ({
+  length: {
+    maximum: maxLength,
+    tooLong: 'must be %{count} characters or less'
+  }
+});
+const validators = {
+  firstName: {...required, ...notTooLong(80)},
+  lastName: {...required, ...notTooLong(80)},
+  currentPosition: {...required, ...notTooLong(255)},
+  organization: {...required, ...notTooLong(255)},
+  currentResearch: required,
+};
+
+export const ProfilePageReact = withUserProfile()
+(class ProfilePageReactComponent extends React.Component<
+  { profileState: {profile: Profile, reload: Function} },
+  { profileEdits: Profile, updating: boolean }
+> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      profileEdits: props.profileState.profile || {},
+      updating: false
+    };
+  }
+
+  componentDidUpdate(prevProps) {
+    const {profileState: {profile}} = this.props;
+
+    if (!fp.isEqual(prevProps.profileState.profile, profile)) {
+      this.setState({profileEdits: profile});
+    }
+  }
+
+  saveProfile() {
+    const {profileState: {reload}} = this.props;
+
+    this.setState({updating: true});
+    profileApi().updateProfile(this.state.profileEdits)
+      .then(() => reload())
+      .then(() => this.setState({updating: false}));
+  }
+
+  render() {
+    const {profileState: {profile}} = this.props;
+    const {profileEdits, updating} = this.state;
+    const {
+      givenName, familyName, currentPosition, organization, areaOfResearch,
+      institutionalAffiliations = []
+    } = profileEdits;
+    const errors = validate({
+      firstName: givenName,
+      lastName: familyName,
+      currentPosition,
+      organization,
+      currentResearch: areaOfResearch
+    }, validators);
+
+    const makeProfileInput = ({title, valueKey, isLong = false, ...props}) => {
+      const errorText = profile && errors && errors[labelSubstitutions[valueKey]];
+
+      const inputProps = {
+        value: fp.get(valueKey, profileEdits) || '',
```

Comment:
fp.get vs `profileEdits[valueKey]`: is this purely style preference or is there something else happening here?

Reply:
This allows deeper access using dot notation, see the institutional affiliations for examples.